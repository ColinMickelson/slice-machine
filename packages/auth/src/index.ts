export { createPrismicAuth } from "./createPrismicAuth";

export { createPrismicAuthMiddleware } from "./createPrismicAuthMiddleware";

export type { PrismicAuth } from "./PrismicAuth";

export type { PrismicUserProfile } from "./types";

export {
	UnauthenticatedError,
	isUnauthenticatedError,
	UnexpectedDataError,
	isUnexpectedDataError,
} from "./errors";
