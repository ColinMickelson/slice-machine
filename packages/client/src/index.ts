import axios, { AxiosError, AxiosPromise, AxiosRequestHeaders } from "axios";
import * as t from "io-ts";

import { UserProfile } from "@slicemachine/core/build/models";
import { SharedSlice } from "@prismicio/types-internal/lib/customtypes/widgets/slices";
import { CustomType } from "@prismicio/types-internal/lib/customtypes/CustomType";

import type { ClientError, UploadParameters } from "./models";
import {
  ApplicationMode,
  getStatus,
  getMessage,
  ApisEndpoints,
  ProductionApisEndpoints,
  StageApisEndpoints,
  AclCreateResult,
} from "./models";

import { getAndValidateResponse } from "./utils";
import { upload } from "./utils/upload";

// exporting models to be used with the Client.
export * from "./models";

// exporting utils so the client calls can be replicated for speficic calls to a consumer.
export * from "./utils";

// exporting the client itself.
export class Client {
  apisEndpoints: ApisEndpoints;
  repository: string | null;
  authenticationToken: string;

  constructor(
    applicationMode: ApplicationMode,
    repository: string | null,
    authenticationToken: string
  ) {
    this.repository = repository;
    this.authenticationToken = authenticationToken;

    if (applicationMode === ApplicationMode.PROD)
      this.apisEndpoints = ProductionApisEndpoints;
    else if (applicationMode === ApplicationMode.STAGE)
      this.apisEndpoints = StageApisEndpoints;
    else {
      // Dev
      this.apisEndpoints = ProductionApisEndpoints;
    }
  }

  // setters to provide flexibility
  updateAuthenticationToken(newToken: string): Client {
    this.authenticationToken = newToken;
    return this
  }

  updateRepository(repository: string | null): Client {
    this.repository = repository;
    return this
  }

  // private methods
  _fetch(
    method: "get" | "post",
    url: string,
    data?: Record<string, unknown>,
    headers?: AxiosRequestHeaders
  ): AxiosPromise {
    return axios({
      method,
      url,
      data,
      headers: {
        Authorization: `Bearer ${this.authenticationToken}`,
        "User-Agent": "slice-machine",
        ...(this.repository ? { repository: this.repository } : {}),
        ...headers,
      },
    });
  }

  _get(url: string): AxiosPromise {
    return this._fetch("get", url);
  }

  _post(url: string, data: Record<string, unknown>): AxiosPromise {
    return this._fetch("post", url, data).catch((error: Error | AxiosError) => {
      const status: number = getStatus(error);
      const message: string = getMessage(error);

      // Making sure the error is typed
      const clientError: ClientError = { status, message };
      return Promise.reject(clientError);
    });
  }

  async validateAuthenticationToken(): Promise<boolean> {
    return this._get(
      `${this.apisEndpoints.Authentication}validate?token=${this.authenticationToken}`
    )
      .then(() => true)
      .catch((error: Error | AxiosError) => {
        const status: number = getStatus(error);
        if (status == 401) return false;

        const message: string = getMessage(error);

        // Making sure the error is typed
        const clientError: ClientError = { status, message };
        return Promise.reject(clientError);
      });
  }

  async refreshAuthenticationToken(): Promise<string> {
    return getAndValidateResponse<string>(
      this._get(
        `${this.apisEndpoints.Authentication}refreshtoken?token=${this.authenticationToken}`
      ),
      "refreshed authentication token",
      t.string
    ).then((newToken) => {
      this.updateAuthenticationToken(newToken);
      return newToken;
    });
  }

  async profile(): Promise<UserProfile> {
    return getAndValidateResponse<UserProfile>(
      this._get(`${this.apisEndpoints.Users}profile`),
      "user profile",
      UserProfile
    );
  }

  async getSlices(): Promise<SharedSlice[]> {
    return getAndValidateResponse<SharedSlice[]>(
      this._get(`${this.apisEndpoints.Models}slices`),
      "shared slices",
      t.array(SharedSlice)
    );
  }

  async getCustomTypes(): Promise<CustomType[]> {
    return getAndValidateResponse<CustomType[]>(
      this._get(`${this.apisEndpoints.Models}customtypes`),
      "custom types",
      t.array(CustomType)
    );
  }

  async insertCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/insert`,
      customType
    );
  }

  async updateCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/update`,
      customType
    );
  }

  async insertSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/insert`, sharedSlice);
  }

  async updateSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/update`, sharedSlice);
  }

  async createImagesAcl(): Promise<AclCreateResult> {
    return getAndValidateResponse<AclCreateResult>(
      this._get(`${this.apisEndpoints.AclProvider}create`),
      "acl",
      AclCreateResult
    );
  }

  async deleteImagesFolderAcl(sliceName: string): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.AclProvider}delete-folder`, {
      sliceName,
    });
  }

  async uploadImageAcl(params: UploadParameters): Promise<number | undefined> {
    return upload(params);
  }
}
