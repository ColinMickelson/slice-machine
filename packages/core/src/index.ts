export * as FileSystem from "./filesystem";
export * as Utils from "./utils";
export * from "./core";
export * as Models from "./models";
export * as Libraries from "./libraries";
export * as Mocks from "./libraries/mocks";
export * from "./auth";
