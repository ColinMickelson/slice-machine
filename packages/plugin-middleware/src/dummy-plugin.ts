import type { SliceMock } from "@slicemachine/core/src/models";

// for testing the dummy, not needed else where
export const name = module.filename;

/* slice template */
export const slice = (name: string): { filename: string; data: string } => ({
  filename: "index.js",
  data: `
const ${name} = () => "foobar"
export default ${name}
`,
});

/* story template */
export const story = (
  path: string,
  title: string,
  mock: SliceMock
): { filename: string; data: string } => ({
  filename: "index.story.js",
  data: "some story",
});

/* how slices are indexed in /slices/index */
export const index = (slices: string[]): { filename: string; data: string } => {
  const imports = slices.map(
    (slice) => `export {default as ${slice}} from "./${slice}"`
  );
  const data = imports.join("\n");
  return {
    filename: "index.js",
    data,
  };
};

// TODO: discuss

/* code snippets, could be hard because currently snippets are created client side */
// export const snippets = (widget: string) => widget

/* simulator maybe make this a different file */
// export const simulator = () => null

/* slice-zone, should we add the slice-zones in the  */
// export sliceZone
