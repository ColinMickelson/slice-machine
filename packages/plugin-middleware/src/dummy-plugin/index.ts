import type { SliceMock } from "@slicemachine/core/src/models";
import type { FieldType } from "@slicemachine/core/src/models/CustomType/fields";

// for testing the dummy, not needed else where
export const name = module.filename;

export const framework = "dummy";
/* slice template */
export const slice = (name: string): { filename: string; data: string } => ({
  filename: "index.js",
  data: `
const ${name} = () => "foobar"
export default ${name}
`,
});

/* story template */
export const story = (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  path: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  title: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  mock: SliceMock
): { filename: string; data: string } => ({
  filename: "index.story.js",
  data: "some story",
});

/* how slices are indexed in /slices/index */
export const index = (slices: string[]): { filename: string; data: string } => {
  const imports = slices.map(
    (slice) => `export {default as ${slice}} from "./${slice}"`
  );
  const data = imports.join("\n");
  return {
    filename: "index.js",
    data,
  };
};

export const snippets = (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  widget: FieldType,
  field: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  useKey?: boolean
): string => {
  // some map for the widget
  return `<div>${field}</div>`;
};

// TODO: discuss

/* code snippets, could be hard because currently snippets are created client side it could be a dynamically imported library, but then we need to figure out which plugin has the hints */
// or make hints a totally sperate thing?, or get the markdown from the server
// export const snippets = (widget: string) => widget

/* simulator maybe make this a different file */
// export const simulator = () => null

/* slice-zone, should we add the slice-zones in the  */
// export sliceZone
