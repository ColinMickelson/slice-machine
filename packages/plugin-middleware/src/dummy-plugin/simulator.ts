/*
Should we add the simulators to the plugins ?
so usage is
export simulator from 'slice-mahcine-next/simulator'
*/

export default {}; // what is the spec for the simulator?
