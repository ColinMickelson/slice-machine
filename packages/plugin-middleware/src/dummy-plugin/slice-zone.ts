/* May aswell add the slice-zones to the framework plugins */
export default "foo";
