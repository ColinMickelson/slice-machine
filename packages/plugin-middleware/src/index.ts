import path from "path";
import type { SliceMock } from "@slicemachine/core/src/models";
import type { FieldType } from "@slicemachine/core/src/models/CustomType/fields";

export type Plugin = {
  slice: (name: string) => { filename: string; data: string };
  story: (
    path: string,
    title: string,
    mock: SliceMock
  ) => { filename: string; data: string };
  index: (slices: string[]) => { filename: string; data: string };
  snippets?: (widget: FieldType, field: string, useKey?: boolean) => string;
  // [key: string]: unknown;
};

export default class PluginContainer {
  plugins: Record<string, Plugin>;
  constructor(paths?: string[]) {
    this.plugins = {};

    this.register = this.register.bind(this);
    this.createSlice = this.createSlice.bind(this);
    this.createStory = this.createStory.bind(this);
    this.createIndex = this.createIndex.bind(this);
    this.createSnippet = this.createSnippet.bind(this);

    if (paths && paths.length) {
      paths.forEach(this.register);
    }
  }

  private _findPluginsWithProp(prop: string): Record<string, Plugin> {
    return Object.entries(this.plugins)
      .filter(([, plugin]) => {
        return Object.prototype.hasOwnProperty.call(plugin, prop);
      })
      .reduce((acc, [name, plugin]) => ({ ...acc, [name]: plugin }), {});
  }

  register(name: string) {
    const pluginPath = name.startsWith(".")
      ? path.resolve(
          process.cwd(),
          path.join.apply(null, name.split(path.posix.sep))
        )
      : name;

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const plugin = require(pluginPath);

    this.plugins[pluginPath] = plugin;
  }

  createSlice(
    sliceName: string
  ): Record<string, { filename: string; data: string }> {
    const slices = this._findPluginsWithProp("slice");
    return Object.entries(slices).reduce((acc, [name, plugin]) => {
      const result = plugin.slice(sliceName);
      return { ...acc, [name]: result };
    }, {});
  }

  createStory(
    path: string,
    title: string,
    mock: SliceMock
  ): Record<string, { filename: string; data: string }> {
    const stories = this._findPluginsWithProp("story");
    return Object.entries(stories).reduce((acc, [name, plugin]) => {
      const result = plugin.story(path, title, mock);
      return { ...acc, [name]: result };
    }, {});
  }

  createIndex(
    stories: string[]
  ): Record<string, { filename: string; data: string }> {
    const indices = this._findPluginsWithProp("index");
    return Object.entries(indices).reduce((acc, [name, plugin]) => {
      const result = plugin.index(stories);
      return { ...acc, [name]: result };
    }, {});
  }

  createSnippet(
    widget: FieldType,
    field: string,
    useKey = false
  ): Record<string, string> {
    const widgets = this._findPluginsWithProp("snippets");
    return Object.entries(widgets).reduce((acc, [name, plugin]) => {
      if (!plugin.snippets) return acc;
      const result = plugin.snippets(widget, field, useKey);
      return { ...acc, [name]: result };
    }, {});
  }
}
