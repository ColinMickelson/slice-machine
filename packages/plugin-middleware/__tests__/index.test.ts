import path from "path";
import PluginMiddleware from "../src";
import * as Dummy from "../src/dummy-plugin";

describe("@slicemachine/plugin-middleware", () => {
  describe("register", () => {
    it("should register a node-module", () => {
      jest.mock("foo", () => ({ bar: null }), { virtual: true });
      const p = new PluginMiddleware(["foo"]);
      expect(p.plugins.foo).toBeDefined();
      expect("bar" in p.plugins.foo).toBeTruthy();
      // expect(p.plugins.foo.bar).toBeNull();
    });

    it("should register a local module", () => {
      jest.mock(
        path.join(process.cwd(), "fake-plugin"),
        () => ({ bar: null }),
        { virtual: true }
      );

      const name = "fake-plugin";
      const p = new PluginMiddleware([`./${name}`]);

      const localPath = path.join(process.cwd(), name);
      expect(p.plugins[localPath]).toBeDefined();
      expect("bar" in p.plugins[localPath]).toBeTruthy();
      // expect(p.plugins[localPath].bar).toBeNull();
    });
  });

  describe("createSlice", () => {
    it("should call the slice method on plugins", () => {
      jest.spyOn(Dummy, "slice");
      const sliceName = "foo";
      const p = new PluginMiddleware([Dummy.name]);
      const result = p.createSlice(sliceName);
      expect(Dummy.slice).toHaveBeenCalled();
      expect(result[Dummy.name].data).toContain(
        `const ${sliceName} = () => "foobar"`
      );
      expect(result[Dummy.name].data).toContain(`export default ${sliceName}`);
      expect(result[Dummy.name].filename).toEqual("index.js");
    });
  });

  describe("createStory", () => {
    it("should call the story method on the plugins", () => {
      jest.spyOn(Dummy, "story");
      const p = new PluginMiddleware([Dummy.name]);
      const result = p.createStory("some/path", "MySlice", []);
      expect(Dummy.story).toBeCalled();
      expect(result[Dummy.name].filename).toEqual("index.story.js");
      // expect(result[Dummy.name].data).toEqual("")
    });
  });

  describe("craeteIndex", () => {
    it("should call the index method on the plugins", () => {
      jest.spyOn(Dummy, "index");
      const p = new PluginMiddleware([Dummy.name]);
      const result = p.createIndex(["foo", "bar", "baz"]);
      expect(Dummy.index).toBeCalled();
      expect(result[Dummy.name].filename).toEqual("index.js");
      expect(result[Dummy.name].data).toContain(
        'export {default as foo} from "./foo"\n'
      );
      expect(result[Dummy.name].data).toContain(
        'export {default as bar} from "./bar"\n'
      );
      expect(result[Dummy.name].data).toContain(
        'export {default as baz} from "./baz"'
      );
    });
  });
});
