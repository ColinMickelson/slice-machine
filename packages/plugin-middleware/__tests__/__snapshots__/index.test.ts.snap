// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@slicemachine/plugin-middleware #createModel should create a default model 1`] = `
Object {
  "data": Object {
    "description": "MySlice",
    "id": "foo_bar",
    "name": "MySlice",
    "type": "SharedSlice",
    "variations": Array [
      Object {
        "description": "MySlice",
        "docURL": "...",
        "id": "default",
        "name": "Default",
        "primary": Object {
          "description": Object {
            "config": Object {
              "label": "Description",
              "placeholder": "A nice description of your product",
              "single": "paragraph",
            },
            "type": "StructuredText",
          },
          "title": Object {
            "config": Object {
              "label": "Title",
              "placeholder": "This is where it all begins...",
              "single": "heading1",
            },
            "type": "StructuredText",
          },
        },
        "version": "sktwi1xtmkfgx8626",
      },
    ],
  },
  "filename": "model.json",
}
`;
